File: forms/otolaryngology-assessment/back-end-with-loco/tests/engine/snot22_grader_test.rs
use otolaryngology_assessment_loco_crate::engine::snot22_grader::{classify_snot22_score, grade};
use otolaryngology_assessment_loco_crate::engine::snot22_rules::all_rules;
use otolaryngology_assessment_loco_crate::engine::types::*;

#[test]
fn empty_case_is_mild_with_zero_score() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.total_score, 0);
    assert_eq!(result.answered_count, 0);
    assert_eq!(result.severity_level, "mild");
}

#[test]
fn classify_bands_match_specification() {
    assert_eq!(classify_snot22_score(0), "mild");
    assert_eq!(classify_snot22_score(7), "mild");
    assert_eq!(classify_snot22_score(8), "moderate");
    assert_eq!(classify_snot22_score(19), "moderate");
    assert_eq!(classify_snot22_score(20), "severe");
    assert_eq!(classify_snot22_score(110), "severe");
}

#[test]
fn answered_items_contribute_to_total() {
    let mut data = AssessmentData::default();
    data.snot22.need_to_blow_nose = Some(5);
    data.snot22.sneezing = Some(3);
    data.snot22.runny_nose = Some(2);
    let result = grade(&data);
    assert_eq!(result.total_score, 10);
    assert_eq!(result.answered_count, 3);
    assert_eq!(result.severity_level, "moderate");
}

#[test]
fn severe_band_when_total_at_or_above_twenty() {
    let mut data = AssessmentData::default();
    // 22 items * 1 = 22 -> severe.
    data.snot22.need_to_blow_nose = Some(1);
    data.snot22.sneezing = Some(1);
    data.snot22.runny_nose = Some(1);
    data.snot22.nasal_blockage = Some(1);
    data.snot22.loss_of_smell_taste = Some(1);
    data.snot22.coughing = Some(1);
    data.snot22.post_nasal_discharge = Some(1);
    data.snot22.thick_nasal_discharge = Some(1);
    data.snot22.ear_fullness = Some(1);
    data.snot22.dizziness = Some(1);
    data.snot22.ear_pain = Some(1);
    data.snot22.facial_pain_pressure = Some(1);
    data.snot22.difficulty_falling_asleep = Some(1);
    data.snot22.waking_up_at_night = Some(1);
    data.snot22.lack_of_good_nights_sleep = Some(1);
    data.snot22.waking_up_tired = Some(1);
    data.snot22.fatigue = Some(1);
    data.snot22.reduced_productivity = Some(1);
    data.snot22.reduced_concentration = Some(1);
    data.snot22.frustrated_restless_irritable = Some(1);
    data.snot22.sad = Some(1);
    data.snot22.embarrassed = Some(1);
    let result = grade(&data);
    assert_eq!(result.total_score, 22);
    assert_eq!(result.answered_count, 22);
    assert_eq!(result.severity_level, "severe");
}

#[test]
fn rule_count_is_22_and_ids_unique() {
    let rules = all_rules();
    assert_eq!(rules.len(), 22, "SNOT-22 should have exactly 22 rules");
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), before, "SNOT-22 rule IDs must be unique");
    // Spot-check the canonical first and last IDs.
    let ids2: Vec<&str> = all_rules().iter().map(|r| r.id).collect();
    assert!(ids2.contains(&"SNOT22-001"));
    assert!(ids2.contains(&"SNOT22-022"));
}

File: forms/otolaryngology-assessment/front-end-with-svelte/src/lib/engine/snot22-grader.test.ts
import { describe, it, expect } from 'vitest';
import { calculateSnot22, classifySnot22Score } from './snot22-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { snot22Rules } from './snot22-rules';
import type { AssessmentData, Snot22Key } from './types';
import { SNOT22_ITEMS } from './types';

/**
 * A blank assessment for tests. Kept local (rather than importing the store's
 * `createDefaultAssessment`) so the engine tests do not pull in SvelteKit's
 * `$app/environment` module, which Vitest does not resolve.
 */
function createDefaultAssessment(): AssessmentData {
	return {
		demographics: { firstName: '', lastName: '', dateOfBirth: '', sex: '', occupation: '' },
		presentingComplaint: {
			earSymptoms: '',
			noseSymptoms: '',
			throatSymptoms: '',
			neckSymptoms: '',
			chiefComplaint: ''
		},
		historyOfPresentIllness: {
			onsetDate: '',
			onsetType: '',
			progression: '',
			laterality: '',
			previousEpisodes: '',
			aggravatingFactors: '',
			relievingFactors: '',
			associatedSymptoms: ''
		},
		pastEntHistory: {
			priorEntSurgery: '',
			priorEntSurgeryDetails: '',
			chronicSinusitis: '',
			allergicRhinitis: '',
			hearingLoss: '',
			tinnitus: '',
			vertigo: '',
			hearingAids: '',
			headNeckCancer: '',
			headNeckRadiotherapy: '',
			smoking: '',
			alcohol: ''
		},
		snot22: {
			needToBlowNose: null,
			sneezing: null,
			runnyNose: null,
			nasalBlockage: null,
			lossOfSmellTaste: null,
			coughing: null,
			postNasalDischarge: null,
			thickNasalDischarge: null,
			earFullness: null,
			dizziness: null,
			earPain: null,
			facialPainPressure: null,
			difficultyFallingAsleep: null,
			wakingUpAtNight: null,
			lackOfGoodNightsSleep: null,
			wakingUpTired: null,
			fatigue: null,
			reducedProductivity: null,
			reducedConcentration: null,
			frustratedRestlessIrritable: null,
			sad: null,
			embarrassed: null
		},
		externalExamination: {
			facialAsymmetry: '',
			facialSwelling: '',
			skinLesions: '',
			externalEarFindings: '',
			externalNoseFindings: '',
			examinationNotes: ''
		},
		otoscopy: {
			right: { tympanicMembrane: '', canal: '', mobility: '' },
			left: { tympanicMembrane: '', canal: '', mobility: '' },
			otoscopyNotes: ''
		},
		anteriorRhinoscopy: {
			right: { septum: '', mucosa: '', polyps: '', discharge: '', turbinateHypertrophy: '' },
			left: { septum: '', mucosa: '', polyps: '', discharge: '', turbinateHypertrophy: '' },
			rhinoscopyNotes: ''
		},
		oropharyngealNeckExamination: {
			oralMucosa: '',
			tonsils: '',
			pharynx: '',
			palateMovement: '',
			cervicalLymphadenopathy: '',
			cervicalLymphadenopathyDetails: '',
			thyroidEnlarged: '',
			neckMass: '',
			neckMassDetails: '',
			examinationNotes: ''
		},
		clinicalImpressionPlan: {
			workingDiagnosis: '',
			differentialDiagnosis: '',
			investigationsRequired: '',
			investigationsDetails: '',
			medicationPrescribed: '',
			medicationDetails: '',
			referralRequired: '',
			referralDetails: '',
			surgeryConsidered: '',
			surgeryDetails: '',
			followUpPlan: '',
			patientEducation: ''
		}
	};
}

/** Set every SNOT-22 item to the same value. */
function fillSnot22(data: AssessmentData, value: number): void {
	for (const { key } of SNOT22_ITEMS) {
		data.snot22[key as Snot22Key] = value;
	}
}

describe('SNOT-22 classification bands', () => {
	it('classifies 0-7 as mild', () => {
		expect(classifySnot22Score(0)).toBe('mild');
		expect(classifySnot22Score(7)).toBe('mild');
	});
	it('classifies 8-19 as moderate', () => {
		expect(classifySnot22Score(8)).toBe('moderate');
		expect(classifySnot22Score(19)).toBe('moderate');
	});
	it('classifies >=20 as severe', () => {
		expect(classifySnot22Score(20)).toBe('severe');
		expect(classifySnot22Score(110)).toBe('severe');
	});
});

describe('SNOT-22 grader', () => {
	it('returns zero total and mild for a blank assessment', () => {
		const data = createDefaultAssessment();
		const result = calculateSnot22(data);
		expect(result.totalScore).toBe(0);
		expect(result.answeredCount).toBe(0);
		expect(result.severityLevel).toBe('mild');
		expect(result.firedRules).toHaveLength(0);
	});

	it('counts only answered items and sums them', () => {
		const data = createDefaultAssessment();
		data.snot22.needToBlowNose = 3;
		data.snot22.nasalBlockage = 2;
		const result = calculateSnot22(data);
		expect(result.answeredCount).toBe(2);
		expect(result.totalScore).toBe(5);
		expect(result.severityLevel).toBe('mild');
	});

	it('computes the maximum 110 total and severe band when all items are 5', () => {
		const data = createDefaultAssessment();
		fillSnot22(data, 5);
		const result = calculateSnot22(data);
		expect(result.totalScore).toBe(110);
		expect(result.answeredCount).toBe(22);
		expect(result.severityLevel).toBe('severe');
		expect(result.firedRules).toHaveLength(22);
	});

	it('treats 0 as an answered item (not blank)', () => {
		const data = createDefaultAssessment();
		data.snot22.sneezing = 0;
		const result = calculateSnot22(data);
		expect(result.answeredCount).toBe(1);
		expect(result.totalScore).toBe(0);
	});

	it('has 22 rules with unique ids', () => {
		expect(snot22Rules).toHaveLength(22);
		const ids = snot22Rules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('SNOT-22 flagged issues detection', () => {
	it('returns no flags for a blank assessment', () => {
		const data = createDefaultAssessment();
		expect(detectAdditionalFlags(data)).toHaveLength(0);
	});

	it('raises an urgent flag for a neck mass', () => {
		const data = createDefaultAssessment();
		data.oropharyngealNeckExamination.neckMass = 'yes';
		const flags = detectAdditionalFlags(data);
		const f = flags.find((x) => x.id === 'FLAG-NECK-001');
		expect(f).toBeDefined();
		expect(f?.priority).toBe('urgent');
	});

	it('raises an urgent flag for sudden-onset hearing loss', () => {
		const data = createDefaultAssessment();
		data.pastEntHistory.hearingLoss = 'yes';
		data.historyOfPresentIllness.onsetType = 'sudden';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-AUD-001' && f.priority === 'urgent')).toBe(true);
	});

	it('flags tympanic membrane perforation per side', () => {
		const data = createDefaultAssessment();
		data.otoscopy.left.tympanicMembrane = 'perforated';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-OTO-PERF-LEFT')).toBe(true);
	});

	it('flags large nasal polyps and asymmetric tonsils', () => {
		const data = createDefaultAssessment();
		data.anteriorRhinoscopy.right.polyps = 'large';
		data.oropharyngealNeckExamination.tonsils = 'asymmetric';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-RHINO-POLYP-RIGHT')).toBe(true);
		expect(flags.some((f) => f.id === 'FLAG-OROPHX-001')).toBe(true);
	});

	it('raises the sleep-domain SNOT-22 flag when the burden is high', () => {
		const data = createDefaultAssessment();
		data.snot22.difficultyFallingAsleep = 3;
		data.snot22.wakingUpAtNight = 3;
		data.snot22.lackOfGoodNightsSleep = 3;
		data.snot22.wakingUpTired = 3;
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-SNOT-SLEEP')).toBe(true);
	});

	it('sorts flags urgent > high > medium > low', () => {
		const data = createDefaultAssessment();
		data.oropharyngealNeckExamination.neckMass = 'yes'; // urgent
		data.externalExamination.skinLesions = 'yes'; // low
		data.pastEntHistory.smoking = 'yes'; // medium
		data.pastEntHistory.headNeckCancer = 'yes'; // high
		const flags = detectAdditionalFlags(data);
		const order = { urgent: 0, high: 1, medium: 2, low: 3 };
		const priorities = flags.map((f) => order[f.priority]);
		const sorted = [...priorities].sort((a, b) => a - b);
		expect(priorities).toEqual(sorted);
	});
});
